ambry-store/src/main/java/com.github.ambry.store/PersistentIndex.java

```diff
@@ -590,15 +605,17 @@ private IndexValue findKey(StoreKey key, FileSpan fileSpan, EnumSet<IndexEntryTy
             }
             logger.trace("Index : {} found value offset {} size {} ttl {}", dataDir, value.getOffset(), value.getSize(),
                 value.getExpiresAtMs());
-            if (types.contains(IndexEntryType.DELETE) && value.isFlagSet(IndexValue.Flags.Delete_Index)) {
+            if (types.contains(IndexEntryType.DELETE) && value.isDelete()) {
+              retCandidate = value;
+              break;
+            } else if (types.contains(IndexEntryType.UNDELETE) && value.isUndelete()) {
               retCandidate = value;
               break;
-            } else if (types.contains(IndexEntryType.TTL_UPDATE) && !value.isFlagSet(IndexValue.Flags.Delete_Index)
-                && value.isFlagSet(IndexValue.Flags.Ttl_Update_Index)) {
+            } else if (types.contains(IndexEntryType.TTL_UPDATE) && !value.isDelete() && !value.isUndelete()
+                && value.isTTLUpdate()) {
```

Comment:
Can you present a case where we do need to check `!value.isDelete() && !value.isUndelete()`? I know the piece of code comes from previous version, I just don't quite understand why we need such additional check.

Reply:
A delete or undelete entry will carry TTL_UPDATE_INDEX flag if there is a ttl_update record comes before it. So here if the search types only contain TTL_UPDATE_INDEX and we are seeing a delete record, then we have to make sure it doesn't get returned, therefore, we have to check `!value.isDelete() && !value.isUndelete()`.